File: source/python/grafo.py
# ==========================================
# ESTRUCTURA DE DATOS: GRAFO (Lista de Adyacencia)
# ==========================================
# Explicación Detallada:
# Un grafo modela relaciones entre objetos. Consiste en 'Vértices' (nodos) y 'Aristas' 
# (conexiones entre ellos).
# En esta implementación usamos "Listas de Adyacencia", que es un diccionario donde 
# la clave es un vértice y su valor es una lista con todos los vértices a los que está conectado.
# Esto es muy eficiente en términos de memoria para grafos dispersos (pocas conexiones).

class Grafo:
    def __init__(self):
        # adj será un diccionario, por ej: {'A': ['B', 'C'], 'B': ['A']}
        self.adj = {}

    # INSERCIÓN (Agregar Vértice y Arista)
    # Proceso: Si los nodos u y v no existen en el diccionario, se crean listas vacías.
    # Como es un grafo NO dirigido, u se conecta con v, y v se conecta con u.
    def agregar_arista(self, u, v):
        if u not in self.adj: self.adj[u] = []
        if v not in self.adj: self.adj[v] = []
        
        self.adj[u].append(v)
        self.adj[v].append(u) # Omitir esta línea para hacer un Grafo Dirigido

    # RECORRIDO / BÚSQUEDA (BFS - Búsqueda en Anchura)
    # Proceso Detallado:
    # 1. Se usa una estructura 'Cola' y un conjunto 'visitados' para evitar ciclos.
    # 2. Agregamos el nodo inicial a la cola y lo marcamos visitado.
    # 3. Mientras la cola no esté vacía, sacamos el nodo y exploramos todos sus "vecinos".
    # 4. Si los vecinos no han sido visitados, los encolamos y los marcamos visitados.
    # El BFS sirve para explorar por "capas" de distancia.
    def bfs(self, inicio):
        visitados = set([inicio])
        cola = [inicio]
        recorrido = []
        
        while cola:
            nodo = cola.pop(0) # Extraer frente de la cola (Dequeue)
            recorrido.append(nodo)
            
            # Explorar vecinos
            for vecino in self.adj.get(nodo, []):
                if vecino not in visitados:
                    visitados.add(vecino)
                    cola.append(vecino)
                    
        print(f"Recorrido BFS comenzando desde '{inicio}':", recorrido)

if __name__ == "__main__":
    g = Grafo()
    # Conexiones (Red de amigos, o carreteras)
    g.agregar_arista('A', 'B')
    g.agregar_arista('A', 'C')
    g.agregar_arista('B', 'D')
    g.agregar_arista('C', 'E')
    
    # Recorrer grafo
    g.bfs('A')

File: source/c/grafo.c
// ==========================================
// ESTRUCTURA DE DATOS: GRAFO (Matriz de Adyacencia)
// ==========================================
// Explicación Detallada:
// Otra forma común de representar un grafo en programación es usando una matriz 2D (Matriz de Adyacencia).
// Las filas y columnas representan los vértices.
// Si hay una arista (conexión) entre el vértice 'i' y el vértice 'j', entonces matriz[i][j] = 1.
// Si no están conectados, es 0.
// Ventaja: Es extremadamente rápido (O(1)) verificar si dos nodos están conectados.
// Desventaja: Desperdicia memoria si hay muchas filas pero pocas conexiones reales (grafos dispersos).

#include <stdio.h>
#define VERTICES 4 // Grafo fijo de 4 nodos para el ejemplo: 0,1,2,3

// Inicializar Grafo
// Proceso: Rellena toda la matriz 2D con 0 indicando que no hay conexiones aún.
void inicializar(int grafo[VERTICES][VERTICES]) {
    for(int i = 0; i < VERTICES; i++) {
        for(int j = 0; j < VERTICES; j++) {
            grafo[i][j] = 0;
        }
    }
}

// INSERCIÓN (Agregar Arista)
// Proceso:
// Asignamos 1 en la intersección de u y v.
// Al ser un Grafo NO dirigido, la matriz es simétrica, por tanto también 
// asignamos 1 en la intersección de v y u.
void agregarArista(int grafo[VERTICES][VERTICES], int u, int v) {
    grafo[u][v] = 1;
    grafo[v][u] = 1; 
}

// RECORRIDO (Impresión de la Matriz)
// Proceso: Imprime la representación tabular visual de las conexiones del grafo.
void recorrer(int grafo[VERTICES][VERTICES]) {
    printf("Matriz de Adyacencia (Nodos 0 al %d):\n", VERTICES-1);
    for(int i = 0; i < VERTICES; i++) {
        printf("Nodo %d -> [ ", i);
        for(int j = 0; j < VERTICES; j++) {
            printf("%d ", grafo[i][j]);
        }
        printf("]\n");
    }
}

int main() {
    int grafo[VERTICES][VERTICES];
    inicializar(grafo);
    
    // Supongamos que 0=A, 1=B, 2=C, 3=D
    agregarArista(grafo, 0, 1); // Conecta 0 con 1
    agregarArista(grafo, 0, 2); // Conecta 0 con 2
    agregarArista(grafo, 1, 3); // Conecta 1 con 3
    
    recorrer(grafo);
    return 0;
}